presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java

```diff
@@ -189,6 +220,39 @@ private void enforceTimeLimits()
         }
     }
 
+    /**
+     *  When cluster reaches max tasks limit and also a single query
+     *  excceds a threshold,  kill this query
+     */
+    private void enforceTaskLimits()
+    {
+        int totalRunningTaskCount = 0;
+        int highestRunningTaskCount = 0;
+        Optional<T> highestRunningTaskQuery = Optional.empty();
+        for (T query : queries.values()) {
+            if (query.isDone()) {
+                continue;
+            }
+            int runningTaskCount = query.getRunningTaskCount();
+            totalRunningTaskCount += runningTaskCount;
+            if (runningTaskCount > highestRunningTaskCount) {
+                highestRunningTaskCount = runningTaskCount;
+                highestRunningTaskQuery = Optional.of(query);
+            }
+        }
+
+        runningTaskCount.set(totalRunningTaskCount);
+
+        if (totalRunningTaskCount > maxTotalRunningTaskCount &&
+                highestRunningTaskCount > maxQueryRunningTaskCount &&
+                highestRunningTaskQuery.isPresent()) {
+            highestRunningTaskQuery.get().fail(new PrestoException(QUERY_HAS_TOO_MANY_STAGES, format(
+                    "Cluster is overloaded with too many tasks (%s), query with highest running tasks is killed (%s). %s",
```

Comment:
Change the message to be more about the query that was killed, similar to the cluster oom error.  Something like:

"Query killed because the cluster is overloaded with too many tasks and this query was running the highest number of tasks"